### DISCUSSION TITLE
[data-importer] Dealing with value replacements

### DISCUSSION THREAD

**Discussion by tjmv** [firefly-iii/firefly-iii#5506] [General]
I'm dealing with a CSV import from my bank (Rabobank NL) and they have a field called *Naam uiteindelijke partij* (*Name final party* for the English readers). It (sometimes) contains a value if a payment processor is being used and it occurs for a few companies in my case. For these somewhat common things I would like to avoid preprocessing the file as a separate step.

I'm not sure if it's intentional or a bug (given the log message), but adding `continue` here would solve the problem for my case: https://github.com/firefly-iii/data-importer/blob/main/app/Services/CSV/Conversion/Routine/LineProcessor.php#L117

This however is not a generic solution for these issues because you cannot rely on the order of the columns. To deal with that you need some kind of weight/priority you can assign to the columns. But I'm not sure it's a good idea to add more things to the interface, it's already quite overwhelming...

Since it's almost impossible to deal with all specialised cases, maybe it's an idea to have user injectable functions as a pre-processing step here https://github.com/firefly-iii/data-importer/blob/main/app/Services/CSV/Conversion/RoutineManager.php#L126. I'm not sure how to best handle this with PHP and I won't spend time on it if it means maintaining a separate fork (and I can just patch the relevant files locally, but that may require *some* work when updating).

Initially I wanted to post an issue for the presumably missing `continue` statement, but since this is all related I made it a broader discussion instead. I won't mind contributing some code that can be of use to some people, but as I said I think it's a waste of time if nobody is going to use it anyway (e.g. when it's easier to just programme a python script and run that every time).

**Comment by JC5** [comment]
Well the missing `continue` statement is a bug, fixed just in time for [0.6.1](https://github.com/firefly-iii/data-importer/releases/tag/0.6.1). 😁 

When it comes to this problem, I'm aware of it. I also bank with Rabobank.

> Since it's almost impossible to deal with all specialised cases

I've decided not to solve this, basically. Experiments with [bank specific code](https://github.com/firefly-iii/csv-importer/tree/3.0.0/app/Services/CSV/Specifics) (user injectable code) show that it's basically my problem to test, maintain and update the code submitted by others. And with banks changing all the time, this quickly becomes impossible to maintain.

Of course it's also possible for people to literally inject their own code, but that's not user friendly enough for other Rabobank users. They'll have to run untrusted code to fix their files, put the files in the right place, it's impossible.

A solution is the Spectre or Nordigen API's, although they don't do much when it comes to data quality. Not because they don't want to, but because it's impossible to maintain 2000+ banks.

I don't want to kill the discussion outright of course, but the data importer will remain, for the time being, in a "what you see is what you get"-state. Something built in Python to clean up the Rabobank's abysmal CSV files would be best, I think.
